### ISSUE TITLE
Rack::Request#host returns unexpected host when the domain starts with a number

### ISSUE THREAD

**Issue by pocke** [rack/rack#1590]
# Problem

`Rack::Request#host` returns unexpected host when the domain starts with a number.

Rack v2.1.2 returns host correctly.

```ruby
gem 'rack', '2.1.2'
require 'rack'

p Rack::Request.new({'HTTP_HOST' => '123foo.example.com'}).host
# => "123foo.example.com"
```

But Rack v2.2.0 (and v2.2.1) returns a part of the host unexpectedly.

```ruby
gem 'rack', '2.2.0'
require 'rack'


p Rack::Request.new({'HTTP_HOST' => '123foo.example.com'}).host
# => "123"
```

I guess the cause is the following pull request. https://github.com/rack/rack/pull/1561
https://github.com/rack/rack/blob/838ce3aee47456573675ba2353f1573158caab32/lib/rack/request.rb#L601-L615
The `AUTHORITY` regexp looks too loose. `123` matches as an `ip4`, but it is not an ip4.

**Comment by ioquatix** [comment]
Yes, that's correct, I guess we need to force it to match the full string. Thanks for the report.

**Comment by ioquatix** [comment]
I thought it would be greedy by default but it wasn't.

Changing `AUTHORITY` to:

```ruby
AUTHORITY = /^
	# The host:
	(?<host>
		# An IPv6 address:
		(\[(?<ip6>.*)\])
		|
		# An IPv4 address:
		(?<ip4>[\d\.]+)
		|
		# A hostname:
		(?<name>[a-zA-Z0-9\.\-]+)
	)
	# The optional port:
	(:(?<port>\d+))?
$/x
```

fixes the issue for me.

**Comment by pocke** [comment]
Thank you for quick fixing!

**Comment by ioquatix** [comment]
Released now in v2.2.2 - thanks for reporting the issue with an easy repro!
